### PR TITLE
Fixes #30295 - Add core events

### PR DESCRIPTION
**Pull request by ofedoren**
@lzap, you might want to review this :) It still needs improvements, but I'd better wait for your notes.

In short, what this includes:
  -  Webhook Targets renamed to Webhooks
  -  Basic UI for Webhook, Payload template addition
  -  Basic API for Webhook, Payload template addition 
  -  Basic permissions
  -  Default template for all webhooks (run `rake db:seed` to automatically add it)
  -  User can subscribe to one event only (haven't tested via API)

What I see needs to be added:
 - [x] - Refactor code to better support only one event for subscribtion
 - [ ] - More macros to deal with events such as `status_changed`
 - [x] - More tests

### CODE REVIEW

File: app/views/foreman_webhooks/payload_templates/webhook_payload_-_payload_default.erb
Comment (lzap) [inline app/views/foreman_webhooks/payload_templates/webhook_payload_-_payload_default.erb:11]: This is quite weird, you essentially create JSON-like Ruby syntax and then let Ruby to convert it :-)

I think what's more readable, safer and user-priendly is just to use ERB to generate the JSON.

File: app/models/payload_template.rb
Comment (lzap) [inline app/models/payload_template.rb:18]: Hmmm can't we simply make Template an abstract model class instead? Quick grep shows that we only create instances in core tests. This is hacky.

File: app/models/payload_template.rb
Comment (lzap) [inline app/models/payload_template.rb:55]: Yes please, definitely useful.

File: app/models/webhook.rb
Comment (lzap) [inline app/models/webhook.rb:39]: The `@>` is costly operation, tablescan on *every* event which occurs in Foreman. I know that for small tables tablescan is usually forced even if there was a join table, but this smells slow. I wish there was a way to add a payload to the subscriber (event id) so every single webhook could subscribe and ID would easily find it. But after looking into Rails codebase it does not appear so.

I think we can probably keep this array approach but if and only if we use Rails.cache for each event fired storing list of webhook ids. Expiration time could be 10 minutes with explicit expiration when a webhook is created, updated or deleted.

File: db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb
Comment (lzap) [inline db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb:5]: I don't have a strong opinion, both work for me.

File: app/controllers/api/v2/webhooks_controller.rb
Comment (lzap) [inline app/controllers/api/v2/webhooks_controller.rb:27]: Do we actually need to have multiple events to be associated with a webhook? I'd rather keep templates simple, I understand that ERB is powerful and probably some things could be generic. To the extreme, one webhook could catch them all - that's probably quite useful use case.

File: app/models/webhook.rb
Comment (lzap) [inline app/models/webhook.rb:15]: What's this allow list for exactly?

**Comment by lzap** [comment]
Overall looks good, please add HTTP METHOD so I can start testing this. I did not review Rails things too much for now.

File: db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb
Comment (timogoebel) [reply db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb:5]: Could `webhooks` mean the history of the actual events?
But I don't have a strong opinion either.

File: lib/foreman_webhooks/engine.rb
Comment (timogoebel) [inline lib/foreman_webhooks/engine.rb:24]: Do we need a migration to remame the permissions?

File: db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb
Comment (lzap) [reply db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb:5]: Good point, do rename it then.

File: app/models/webhook.rb
Comment (lzap) [reply app/models/webhook.rb:15]: Wait isn't there a method which returns this list? Or why we need it here?

Anyway I agree this should be defined in core so other plugins can use it and modify it.

File: app/controllers/api/v2/webhooks_controller.rb
Comment (ofedoren) [reply app/controllers/api/v2/webhooks_controller.rb:29]: For early testing purposes I've added a possibility to set content-type and http method via UI/API.

Now I'm going to add the rest in a new commit. The bad thing about is that it will be a bit messy with the database: you'll need to rollback the latest migration, then checkout and then migrate. For the sake of keeping addition of new columns in one migration. Although I don't mind if it will be one or more migrations.

**Review by ofedoren — COMMENTED** [review]
Thanks, @lzap and @timogoebel for reviews!

I've tried to address all the comments. Hopefully it makes more sense now.

File: app/controllers/api/v2/webhooks_controller.rb
Comment (timogoebel) [inline app/controllers/api/v2/webhooks_controller.rb:28]: Why did you change `events` to `event`?

File: app/models/webhook.rb
Comment (timogoebel) [inline app/models/webhook.rb:25]: Nit: Do we want to improve the validator here? Does this give a meaningful message?

File: app/controllers/api/v2/webhooks_controller.rb
Comment (lzap) [reply app/controllers/api/v2/webhooks_controller.rb:29]: Yeah rollbacking is something we can't avoid at some point.

File: app/views/foreman_webhooks/payload_templates/webhook_payload_-_payload_default.erb
Comment (lzap) [reply app/views/foreman_webhooks/payload_templates/webhook_payload_-_payload_default.erb:11]: Okay, might be actually useful. Let's see this in practice.

File: app/models/webhook.rb
Comment (lzap) [reply app/models/webhook.rb:42]: Before creating an optimilization ticket, let's solve the big question. Do we want to have ability to subscribe to multiple events for a single webhook? I only see one particular usecase - to create some kind of one-catch-them-all example for debugging, but overall business-wise I don't see use cases. @timogoebel what's your take on this? Can we refactor the array to a simple string?

Comment (lzap) [reply app/models/webhook.rb:42]: Just not to forget: https://projects.theforeman.org/issues/30796

**Comment by lzap** [comment]
Staring testing this, please do not modify migrations from now on unless it's a bug. Tell me if you do so I can rollback it, thanks.

**Comment by lzap** [comment]
I am getting `undefined method `delete_suffix' for nil:NilClass` when trying to access the page. A helper that you did not push?

**Comment by lzap** [comment]
Oh wait it's a Ruby method. This fixes it:

```diff
diff --git a/app/views/webhooks/_form.html.erb b/app/views/webhooks/_form.html.erb
index 5ccbf63..4988d43 100644
--- a/app/views/webhooks/_form.html.erb
+++ b/app/views/webhooks/_form.html.erb
@@ -12,9 +12,9 @@
 
   <%= render('templates', f: f) %>
 
-  <%= selectable_f f, :event, Webhook.available_events.sort.map { |e| e.delete_suffix(Webhook::EVENT_POSTFIX) },
+  <%= selectable_f f, :event, Webhook.available_events.sort.map { |e| e&.delete_suffix(Webhook::EVENT_POSTFIX) },
                    { include_blank: false,
-                     selected: @webhook.event.delete_suffix(Webhook::EVENT_POSTFIX) },
+                     selected: @webhook.event&.delete_suffix(Webhook::EVENT_POSTFIX) },
                    { label: _('Subscribe to'),
                      label_help: _(f.object.try(:description)),
                      required: true } %>
```

File: lib/foreman_webhooks/engine.rb
Comment (lzap) [inline lib/foreman_webhooks/engine.rb:59]: To be more consistent with the item "Webhooks" above, I suggest to rename this to "Webhook Templates". Payload is too generic term.

**Comment by lzap** [comment]
Now I am getting `undefined method '#id' for NilClass::Jail (NilClass)` when rendering the template. The reason for that is in the default template you have `@object` while it should be `@payload.object`.

Here is what the output looks like with shellhooks plugin. For some reason tho shellhooks is getting the request twice and not once. Investigating.

```
2020-09-09T13:44:48 500e3d38 [I] Started POST /print_body 
2020-09-09T13:44:48 500e3d38 [D] Headers: {"HTTP_ACCEPT_ENCODING"=>"gzip;q=1.0,deflate;q=0.6,identity;q=0.3", "HTTP_ACCEPT"=>"*/*", "HTTP_USER_AGENT"=>"Ruby", "HTTP_CONNECTION"=>"close", "HTTP_HOST"=>"localhost:8448", "HTTP_VERSION"=>"HTTP/1.1"}
2020-09-09T13:44:48 500e3d38 [D] Headers: {"HTTP_ACCEPT_ENCODING"=>"gzip;q=1.0,deflate;q=0.6,identity;q=0.3", "HTTP_ACCEPT"=>"*/*", "HTTP_USER_AGENT"=>"Ruby", "HTTP_CONNECTION"=>"close", "HTTP_HOST"=>"localhost:8448", "HTTP_VERSION"=>"HTTP/1.1"}
2020-09-09T13:44:48 500e3d38 [D] Body: 
{"object":"nil","event_name":"domain_updated.event.foreman","webhook_id":1,"payload":"{:object=\u003e#\u003cDomain id: 1, name: \"dummy.lan\", fullname: \"\", created_at: \"2020-08-26 12:01:10\", updated_at: \"2020-09-09 11:38:17\", dns_id: nil\u003e, :context=\u003e{:remote_ip=\u003e\"::1\", :request=\u003e\"51e31868-aa7e-4c45-a018-ee5e98ed882a\", :session=\u003e\"3cecece3-994d-4443-ab9e-b59eb0010ee9\", :user_login=\u003e\"admin\", :user_admin=\u003etrue}}","context":{"remote_ip":"::1","request":"51e31868-aa7e-4c45-a018-ee5e98ed882a","session":"3cecece3-994d-4443-ab9e-b59eb0010ee9","user_login":"admin","user_admin":true}}

2020-09-09T13:44:48 500e3d38 [D] Body: 
{"object":"nil","event_name":"domain_updated.event.foreman","webhook_id":1,"payload":"{:object=\u003e#\u003cDomain id: 1, name: \"dummy.lan\", fullname: \"\", created_at: \"2020-08-26 12:01:10\", updated_at: \"2020-09-09 11:38:17\", dns_id: nil\u003e, :context=\u003e{:remote_ip=\u003e\"::1\", :request=\u003e\"51e31868-aa7e-4c45-a018-ee5e98ed882a\", :session=\u003e\"3cecece3-994d-4443-ab9e-b59eb0010ee9\", :user_login=\u003e\"admin\", :user_admin=\u003etrue}}","context":{"remote_ip":"::1","request":"51e31868-aa7e-4c45-a018-ee5e98ed882a","session":"3cecece3-994d-4443-ab9e-b59eb0010ee9","user_login":"admin","user_admin":true}}

2020-09-09T13:44:48 7a2f5413 [I] Started POST /shellhook/print_body 
2020-09-09T13:44:48 7a2f5413 [D] Headers: {"HTTP_ACCEPT_ENCODING"=>"gzip;q=1.0,deflate;q=0.6,identity;q=0.3", "HTTP_ACCEPT"=>"*/*", "HTTP_USER_AGENT"=>"Ruby", "HTTP_CONNECTION"=>"close", "HTTP_HOST"=>"localhost:8448", "HTTP_VERSION"=>"HTTP/1.1"}
2020-09-09T13:44:48 7a2f5413 [D] Body: 
{"object":"nil","event_name":"domain_updated.event.foreman","webhook_id":1,"payload":"{:object=\u003e#\u003cDomain id: 1, name: \"dummy.lan\", fullname: \"\", created_at: \"2020-08-26 12:01:10\", updated_at: \"2020-09-09 11:38:17\", dns_id: nil\u003e, :context=\u003e{:remote_ip=\u003e\"::1\", :request=\u003e\"51e31868-aa7e-4c45-a018-ee5e98ed882a\", :session=\u003e\"3cecece3-994d-4443-ab9e-b59eb0010ee9\", :user_login=\u003e\"admin\", :user_admin=\u003etrue}}","context":{"remote_ip":"::1","request":"51e31868-aa7e-4c45-a018-ee5e98ed882a","session":"3cecece3-994d-4443-ab9e-b59eb0010ee9","user_login":"admin","user_admin":true}}

2020-09-09T13:44:48 7a2f5413 [I] [47559] Started task ["/home/lzap/work/smart_proxy_shellhooks/examples/print_body"]
2020-09-09T13:44:48 7a2f5413 [I] Finished POST /shellhook/print_body with 200 (5.88 ms)
2020-09-09T13:44:48 7a2f5413 [D] [47559] 

2020-09-09T13:44:48 7a2f5413 [I] Finished POST /print_body with 200 (7.16 ms)
2020-09-09T13:44:48 7a2f5413 [D] [47559] {"object":"nil","event_name":"domain_updated.event.foreman","webhook_id":1,"payload":"{:object=\u003e#\u003cDomain id: 1, name: \"dummy.lan\", fullname: \"\", created_at: \"2020-08-26 12:01:10\", updated_at: \"2020-09-09 11:38:17\", dns_id: nil\u003e, :context=\u003e{:remote_ip=\u003e\"::1\", :request=\u003e\"51e31868-aa7e-4c45-a018-ee5e98ed882a\", :session=\u003e\"3cecece3-994d-4443-ab9e-b59eb0010ee9\", :user_login=\u003e\"admin\", :user_admin=\u003etrue}}","context":{"remote_ip":"::1","request":"51e31868-aa7e-4c45-a018-ee5e98ed882a","session":"3cecece3-994d-4443-ab9e-b59eb0010ee9","user_login":"admin","user_admin":true}}

2020-09-09T13:44:48 7a2f5413 [I] Finished POST /print_body with 200 (7.16 ms)
```

**Comment by lzap** [comment]
Observation - these logging lines are huge, please turn them into debug:

```
2020-09-09T13:50:21 [I|app|a58b5d81] Enqueued ForemanWebhooks::DeliverWebhookJob (Job ID: 2636c327-c50e-4632-a155-c98d6bd512e7) to Dynflow(default) with arguments: {:event_name=>"domain_updated.event.foreman", :payload=>"\n{\"object\":\"nil\",\"event_name\":\"domain_updated.event.foreman\",\"webhook_id\":1,\"payload\":\"{:object=\\u003e#\\u003cDomain id: 1, name: \\\"dummy.lan\\\", fullname: \\\"\\\", created_at: \\\"2020-08-26 12:01:10\\\", updated_at: \\\"2020-09-09 11:38:17\\\", dns_id: nil\\u003e, :context=\\u003e{:remote_ip=\\u003e\\\"::1\\\", :request=\\u003e\\\"a58b5d81-89c1-45bf-95d0-74c99e1248d3\\\", :session=\\u003e\\\"3cecece3-994d-4443-ab9e-b59eb0010ee9\\\", :user_login=\\u003e\\\"admin\\\", :user_admin=\\u003etrue}}\",\"context\":{\"remote_ip\":\"::1\",\"request\":\"a58b5d81-89c1-45bf-95d0-74c99e1248d3\",\"session\":\"3cecece3-994d-4443-ab9e-b59eb0010ee9\",\"user_login\":\"admin\",\"user_admin\":true}}\n", :webhook_id=>1}

and

2020-09-09T13:50:21 [I|app|a58b5d81] Performing ForemanWebhooks::DeliverWebhookJob (Job ID: 2636c327-c50e-4632-a155-c98d6bd512e7) from Dynflow(default) enqueued at 2020-09-09T11:50:21Z with arguments: {:event_name=>"domain_updated.event.foreman", :payload=>"\n{\"object\":\"nil\",\"event_name\":\"domain_updated.event.foreman\",\"webhook_id\":1,\"payload\":\"{:object=\\u003e#\\u003cDomain id: 1, name: \\\"dummy.lan\\\", fullname: \\\"\\\", created_at: \\\"2020-08-26 12:01:10\\\", updated_at: \\\"2020-09-09 11:38:17\\\", dns_id: nil\\u003e, :context=\\u003e{:remote_ip=\\u003e\\\"::1\\\", :request=\\u003e\\\"a58b5d81-89c1-45bf-95d0-74c99e1248d3\\\", :session=\\u003e\\\"3cecece3-994d-4443-ab9e-b59eb0010ee9\\\", :user_login=\\u003e\\\"admin\\\", :user_admin=\\u003etrue}}\",\"context\":{\"remote_ip\":\"::1\",\"request\":\"a58b5d81-89c1-45bf-95d0-74c99e1248d3\",\"session\":\"3cecece3-994d-4443-ab9e-b59eb0010ee9\",\"user_login\":\"admin\",\"user_admin\":true}}\n", :webhook_id=>1}
```

**Comment by lzap** [comment]
Ok the dual/triple logging is a bug in smart proxy. It gets delivered just once. Any comments @timogoebel ? We'd like to keep PRs rather small so I suggest to merge this one once tests are improved and continue in smaller steps.

File: app/models/webhook.rb
Comment (lzap) [inline app/models/webhook.rb:77]: I wonder if we just passed `payload.object` as `@object`.

**Comment by timogoebel** [comment]
> Any comments @timogoebel ? We'd like to keep PRs rather small so I suggest to merge this one once tests are improved and continue in smaller steps.

Feel free to merge, we can and should iterate on this.

**Comment by lzap** [comment]
I run into interesting bug, when Rails are reloaded I am getting this strange error:

```
2020-09-11T15:05:56 [I|app|7ef98646] ForemanWebhooks::EventSubscriber: host_updated.event.foreman event received
2020-09-11T15:05:56 [W|app|7ef98646] Action failed
2020-09-11T15:05:56 [I|app|7ef98646] Backtrace for 'Action failed' error (ArgumentError): A copy of ForemanWebhooks::EventSubscriber has been removed from the module tree but is still active!
 7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/dependencies.rb:529:in `load_missing_constant'
 7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/active_support.rb:61:in `block in load_missing_constant'
 7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/active_support.rb:17:in `allow_bootsnap_retry'
 7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/active_support.rb:60:in `load_missing_constant'
 7ef98646 | /home/lzap/work/foreman_hooks/lib/foreman_hooks/as_dependencies_hook.rb:4:in `load_missing_constant'
 7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/dependencies.rb:214:in `const_missing'
 7ef98646 | /home/lzap/work/foreman_webhooks/app/subscribers/foreman_webhooks/event_subscriber.rb:6:in `call'
 7ef98646 | /home/lzap/work/foreman/app/subscribers/foreman/base_subscriber.rb:6:in `call'
 7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/notificat
...
```

**Review by ofedoren — COMMENTED** [review]
@lzap, updated. More fields for webhook per RFC, basic auth support.

**Important**: before applying this update, please don't forget to follow these steps:
- Don't checkout new commit to local branch yet. 
- Ideally remove all the data related to the current state of the plugin, e.g. templates, webhooks, old permissions. I didn't create a migration for this, since this is not useful after this PR is merged, you can rename them inspiring by the existing migration.
 - Rollback the migration with template addition to the webhooks table: `bundle exec rails db:migrate:down VERSION=20200831194514`
 - Rollback the migration with 2 new columns for the webhooks table: `bundle exec rails db:migrate:down VERSION=20200908004234`
 - Now you can checkout the new commit safe.
 - Apply new migration: `bundle exec rails db:migrate` (or one by one with commands mentioned above. Please, use `db:migrate:up` with the same version instead of `down`)

> Observation - these logging lines are huge, please turn them into debug:
> 
> ```
> ...
> ```
Those are coming from Dynflow if I'm not mistaken, not from here.

**Comment by ofedoren** [comment]
> I run into interesting bug, when Rails are reloaded I am getting this strange error:
> 
> ```
> 2020-09-11T15:05:56 [I|app|7ef98646] ForemanWebhooks::EventSubscriber: host_updated.event.foreman event received
> 2020-09-11T15:05:56 [W|app|7ef98646] Action failed
> 2020-09-11T15:05:56 [I|app|7ef98646] Backtrace for 'Action failed' error (ArgumentError): A copy of ForemanWebhooks::EventSubscriber has been removed from the module tree but is still active!
>  7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/dependencies.rb:529:in `load_missing_constant'
>  7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/active_support.rb:61:in `block in load_missing_constant'
>  7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/active_support.rb:17:in `allow_bootsnap_retry'
>  7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/active_support.rb:60:in `load_missing_constant'
>  7ef98646 | /home/lzap/work/foreman_hooks/lib/foreman_hooks/as_dependencies_hook.rb:4:in `load_missing_constant'
>  7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/dependencies.rb:214:in `const_missing'
>  7ef98646 | /home/lzap/work/foreman_webhooks/app/subscribers/foreman_webhooks/event_subscriber.rb:6:in `call'
>  7ef98646 | /home/lzap/work/foreman/app/subscribers/foreman/base_subscriber.rb:6:in `call'
>  7ef98646 | /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/notificat
> ...
> ```

I've noticed that I was getting such errors way earlier for other plugins.. Simple reload fixed that.

File: db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb
Comment (lzap) [reply db/migrate/20200831194208_rename_webhook_targets_to_webhooks.rb:5]: So we have an agreement to rename.

File: app/controllers/api/v2/webhooks_controller.rb
Comment (lzap) [reply app/controllers/api/v2/webhooks_controller.rb:28]: I thought we are not changing this in this PR.

File: app/lib/foreman_webhooks/renderer/scope/webhook_template.rb
Comment (lzap) [inline app/lib/foreman_webhooks/renderer/scope/webhook_template.rb:32]: Nitpick: Since macros are global by default, this should be probably something like `hash_to_json` or something like that. We can change later.

File: app/models/webhook.rb
Comment (lzap) [reply app/models/webhook.rb:16]: I don't like filtering in the codebase and I tend to remove this completely, show them all but improve our UI so it's easier for users to pick items: https://community.theforeman.org/t/ux-dropdowns-with-many-items/20406

Again, let's keep it this way for now.

**Review by lzap — APPROVED** [review]
Works fine, let's iterate more quickly with smaller PRs from now on! :-)

File: app/lib/foreman_webhooks/renderer/scope/webhook_template.rb
Comment (timogoebel) [reply app/lib/foreman_webhooks/renderer/scope/webhook_template.rb:32]: Actually, that's why we have the concept of scopes: So we can add macros that are only allowed in specific scenarios.

Note, you can use ERB in parameters. There might be macros you don't want to allow for all users.

**Comment by timogoebel** [comment]
Thanks @ofedoren and @lzap. Happy to see the great progress her!

**Comment by lzap** [comment]
Next up: CLI and React UI. Then refinements and demo in three weeks.

**Comment by timogoebel** [comment]
> Next up: CLI and React UI. Then refinements and demo in three weeks.

Where do we track the open issues? Maybe @kamils-iRonin can help with one or two.

**Comment by lzap** [comment]
Here is our tracking ticket: https://projects.theforeman.org/issues/30290

**Comment by lzap** [comment]
@kamils-iRonin if you need a service to test this against use the tiny https://github.com/theforeman/smart_proxy_shellhooks

**Comment by lzap** [comment]
One in particular is bugging me all the time and I could not figure it out yet: https://projects.theforeman.org/issues/31075 if you have an idea @kamils-iRonin ...

**Pull request merged**
